In [103]:
import pandas as pd

In [104]:
data=pd.read_csv("/content/train.csv")

In [105]:
data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

## A) Looking for the missing values

In [106]:
data.isna().sum()

Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64

In [107]:
d2= pd.DataFrame(data)

## B) Removing the units from Mileage, Engine, Power and New_Price and converting them into float to keep in numerical values

In [108]:
d2['Mileage'] = d2['Mileage'].str.extract('(\d+\.\d+)').astype(float)

d2['Engine'] = d2['Engine'].str.replace(' CC', '').astype(float)

d2['Power'] = d2['Power'].str.extract('(\d+\.\d+)').astype(float)

d2['New_Price'] = d2['New_Price'].str.extract('(\d+\.\d+)').astype(float)

d2.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type  Mileage  Engine  \
0              41000    Diesel       Manual      First    19.67  1582.0   
1              46000    Petrol       Manual      First      NaN  1199.0   
2              87000    Diesel       Manual      First    20.77  1248.0   
3              40670    Diesel    Automatic     Second    15.20  1968.0   
4              86999    Diesel       Manual      First    23.08  1461.0   

    Power  Seats  New_Price  Price  
0  126.20    5.0        NaN  12.50  
1   88.70    5.0       8.61   4.50  
2   88.76    7.0        NaN   6.00  
3  140.80    5.0        NaN  17.74  
4   63.10    5.0        NaN   3.50

## (A) Filling in missing values im Mileage, Engine, Power,Seats with median

In [109]:
d2['Mileage'].fillna(d2['Mileage'].median(),inplace=True)
d2['Engine'].fillna(d2['Engine'].median(),inplace=True)
d2['Power'].fillna(d2['Power'].median(),inplace=True)

d2['Seats'].fillna(d2['Seats'].mode()[0],inplace=True)
d2=d2.dropna()
d2.head()

Unnamed: 0                         Name Location  Year  Kilometers_Driven  \
1            2                 Honda Jazz V  Chennai  2011              46000   
8           10             Maruti Ciaz Zeta    Kochi  2018              25692   
13          15  Mitsubishi Pajero Sport 4X4    Delhi  2014             110000   
18          20            BMW 3 Series 320d    Kochi  2014              32982   
26          28          Honda WRV i-VTEC VX    Kochi  2018              37430   

   Fuel_Type Transmission Owner_Type  Mileage  Engine   Power  Seats  \
1     Petrol       Manual      First    18.19  1199.0   88.70    5.0   
8     Petrol       Manual      First    21.56  1462.0  103.25    5.0   
13    Diesel       Manual      First    13.50  2477.0  175.56    7.0   
18    Diesel    Automatic      First    22.69  1995.0   88.80    5.0   
26    Petrol       Manual      First    17.50  1199.0   88.70    5.0   

    New_Price  Price  
1        8.61   4.50  
8       10.65   9.95  
13      32.01  15.00  
18      47.87  18.55  
26      10.57   9.90

In [110]:
output_csv_file = 'preprocessed_data.csv'
d2.to_csv(output_csv_file, index=False)

Determine which columns in this job have missing values: Mileage (2), Engine (36), Power (36), Seats (38) and New Price (5032). Since there are more missing values in this instance, the New Price column was removed from the data frame. As the number of missing information for mileage, engine, power, and seats decreases assigned a "median" value to them. When compared to imputation using mean, the robust strategy of imputing missing values using the median is less susceptible to outliers and works well for a variety of data distributions.

## C) Changing the categorical variables to numerical value

In [111]:
d1=pd.read_csv("/content/preprocessed_data.csv")

In [112]:
d1.isna().sum()

Unnamed: 0           0
Name                 0
Location             0
Year                 0
Kilometers_Driven    0
Fuel_Type            0
Transmission         0
Owner_Type           0
Mileage              0
Engine               0
Power                0
Seats                0
New_Price            0
Price                0
dtype: int64

In [ ]:
#categorical variables: "Fuel_Type," "Transmission," and "Owner_Type."

In [113]:
d1['Fuel_Type'].unique()

array(['Petrol', 'Diesel', 'Electric'], dtype=object)

In [114]:
d1['Transmission'].unique()

array(['Manual', 'Automatic'], dtype=object)

In [115]:
d1['Owner_Type'].unique()

array(['First', 'Second', 'Third'], dtype=object)

In [116]:
#To work with our categorical variables, we are converting them into numerical data using one-hot encoding.
d1['Fuel_Type'].replace({'Diesel': 0, 'Petrol': 1, 'Electric': 2}, inplace=True)
d1['Transmission'].replace({'Manual': 0, 'Automatic': 1}, inplace=True)
d1['Owner_Type'].replace({'First': 1, 'Second': 2, 'Third': 3, 'Fourth & Above': 4}, inplace=True)

In [117]:
d1.head()

Unnamed: 0                         Name Location  Year  Kilometers_Driven  \
0           2                 Honda Jazz V  Chennai  2011              46000   
1          10             Maruti Ciaz Zeta    Kochi  2018              25692   
2          15  Mitsubishi Pajero Sport 4X4    Delhi  2014             110000   
3          20            BMW 3 Series 320d    Kochi  2014              32982   
4          28          Honda WRV i-VTEC VX    Kochi  2018              37430   

   Fuel_Type  Transmission  Owner_Type  Mileage  Engine   Power  Seats  \
0          1             0           1    18.19  1199.0   88.70    5.0   
1          1             0           1    21.56  1462.0  103.25    5.0   
2          0             0           1    13.50  2477.0  175.56    7.0   
3          0             1           1    22.69  1995.0   88.80    5.0   
4          1             0           1    17.50  1199.0   88.70    5.0   

   New_Price  Price  
0       8.61   4.50  
1      10.65   9.95  
2      32.01  15.00  
3      47.87  18.55  
4      10.57   9.90

In [118]:
#Save the encoded data in separate csv file

In [119]:
output_csv_file = 'encoded_data.csv'
d1.to_csv(output_csv_file, index=False)

## D)We are adding a new column called "CURRENT AGE OF THE CAR" to our dataset

In [120]:
from datetime import datetime

In [121]:
current_year = datetime.now().year

d1['Current_Age'] = current_year - d1['Year']

In [122]:
#Save the encoded data in separate csv file
output_csv_file = 'newColumnData.csv'
d1.to_csv(output_csv_file, index=False)

In [123]:
d3=pd.read_csv("/content/newColumnData.csv")
d3.head()

Unnamed: 0                         Name Location  Year  Kilometers_Driven  \
0           2                 Honda Jazz V  Chennai  2011              46000   
1          10             Maruti Ciaz Zeta    Kochi  2018              25692   
2          15  Mitsubishi Pajero Sport 4X4    Delhi  2014             110000   
3          20            BMW 3 Series 320d    Kochi  2014              32982   
4          28          Honda WRV i-VTEC VX    Kochi  2018              37430   

   Fuel_Type  Transmission  Owner_Type  Mileage  Engine   Power  Seats  \
0          1             0           1    18.19  1199.0   88.70    5.0   
1          1             0           1    21.56  1462.0  103.25    5.0   
2          0             0           1    13.50  2477.0  175.56    7.0   
3          0             1           1    22.69  1995.0   88.80    5.0   
4          1             0           1    17.50  1199.0   88.70    5.0   

   New_Price  Price  Current_Age  
0       8.61   4.50           13  
1      10.65   9.95            6  
2      32.01  15.00           10  
3      47.87  18.55           10  
4      10.57   9.90            6

## E) Perform Operations

## Select operations

In [124]:
selected_columns = d3[['Current_Age','Mileage', 'Price']]
print("Selected DataFrame:")
print(selected_columns.head())


Selected DataFrame:
   Current_Age  Mileage  Price
0           13    18.19   4.50
1            6    21.56   9.95
2           10    13.50  15.00
3           10    22.69  18.55
4            6    17.50   9.90


## Filtering the data

In [125]:
d3['Mileage'] = pd.to_numeric(d3['Mileage'], errors='coerce')
filtered_data = d3[d3['Mileage'] > 20]
print("filtered_dataframe")
print(filtered_data.head())


filtered_dataframe
    Unnamed: 0                             Name Location  Year  \
1           10                 Maruti Ciaz Zeta    Kochi  2018   
3           20                BMW 3 Series 320d    Kochi  2014   
6           36    Maruti Vitara Brezza ZDi Plus    Kochi  2018   
11          69  Hyundai Creta 1.4 E Plus Diesel   Mumbai  2018   
17         113    BMW 3 Series 320d Luxury Line   Mumbai  2015   

    Kilometers_Driven  Fuel_Type  Transmission  Owner_Type  Mileage  Engine  \
1               25692          1             0           1    21.56  1462.0   
3               32982          0             1           1    22.69  1995.0   
6               50075          0             0           1    24.30  1248.0   
11              13000          0             0           1    22.10  1396.0   
17              56087          0             1           1    22.69  1995.0   

     Power  Seats  New_Price  Price  Current_Age  
1   103.25    5.0      10.65   9.95            6  
3    88

#Renaming the columns

In [126]:
d3.rename(columns={'Year': 'original_Year'}, inplace=True)

d3

Unnamed: 0                           Name Location  original_Year  \
0             2                   Honda Jazz V  Chennai           2011   
1            10               Maruti Ciaz Zeta    Kochi           2018   
2            15    Mitsubishi Pajero Sport 4X4    Delhi           2014   
3            20              BMW 3 Series 320d    Kochi           2014   
4            28            Honda WRV i-VTEC VX    Kochi           2018   
..          ...                            ...      ...            ...   
796        5999          Tata Bolt Revotron XT  Chennai           2016   
797        6002  Volkswagen Vento 1.6 Highline   Mumbai           2011   
798        6005       Maruti Vitara Brezza VDi     Pune           2016   
799        6010           Honda Brio 1.2 VX MT    Delhi           2013   
800        6014               Maruti Swift VDI    Delhi           2014   

     Kilometers_Driven  Fuel_Type  Transmission  Owner_Type  Mileage  Engine  \
0                46000          1             0           1    18.19  1199.0   
1                25692          1             0           1    21.56  1462.0   
2               110000          0             0           1    13.50  2477.0   
3                32982          0             1           1    22.69  1995.0   
4                37430          1             0           1    17.50  1199.0   
..                 ...        ...           ...         ...      ...     ...   
796              10000          1             0           1    17.57  1193.0   
797              38000          1             0           1    16.09  1598.0   
798              37208          0             0           1    24.30  1248.0   
799              33746          1             0           1    18.50  1198.0   
800              27365          0             0           1    28.40  1248.0   

      Power  Seats  New_Price  Price  Current_Age  
0     88.70    5.0       8.61   4.50           13  
1    103.25    5.0      10.65   9.95            6  
2    175.56    7.0      32.01  15.00           10  
3     88.80    5.0      47.87  18.55           10  
4     88.70    5.0      10.57   9.90            6  
..      ...    ...        ...    ...          ...  
796   88.70    5.0       7.77   4.00            8  
797  103.50    5.0      11.91   3.25           13  
798   88.50    5.0       9.93   7.43            8  
799   86.80    5.0       6.63   3.20           11  
800   88.80    5.0       7.88   4.75           10  

[801 rows x 15 columns]

#Mutating the column

In [127]:
current_year = 2024
d3['Age'] = current_year - d3['original_Year']
d3


Unnamed: 0                           Name Location  original_Year  \
0             2                   Honda Jazz V  Chennai           2011   
1            10               Maruti Ciaz Zeta    Kochi           2018   
2            15    Mitsubishi Pajero Sport 4X4    Delhi           2014   
3            20              BMW 3 Series 320d    Kochi           2014   
4            28            Honda WRV i-VTEC VX    Kochi           2018   
..          ...                            ...      ...            ...   
796        5999          Tata Bolt Revotron XT  Chennai           2016   
797        6002  Volkswagen Vento 1.6 Highline   Mumbai           2011   
798        6005       Maruti Vitara Brezza VDi     Pune           2016   
799        6010           Honda Brio 1.2 VX MT    Delhi           2013   
800        6014               Maruti Swift VDI    Delhi           2014   

     Kilometers_Driven  Fuel_Type  Transmission  Owner_Type  Mileage  Engine  \
0                46000          1             0           1    18.19  1199.0   
1                25692          1             0           1    21.56  1462.0   
2               110000          0             0           1    13.50  2477.0   
3                32982          0             1           1    22.69  1995.0   
4                37430          1             0           1    17.50  1199.0   
..                 ...        ...           ...         ...      ...     ...   
796              10000          1             0           1    17.57  1193.0   
797              38000          1             0           1    16.09  1598.0   
798              37208          0             0           1    24.30  1248.0   
799              33746          1             0           1    18.50  1198.0   
800              27365          0             0           1    28.40  1248.0   

      Power  Seats  New_Price  Price  Current_Age  Age  
0     88.70    5.0       8.61   4.50           13   13  
1    103.25    5.0      10.65   9.95            6    6  
2    175.56    7.0      32.01  15.00           10   10  
3     88.80    5.0      47.87  18.55           10   10  
4     88.70    5.0      10.57   9.90            6    6  
..      ...    ...        ...    ...          ...  ...  
796   88.70    5.0       7.77   4.00            8    8  
797  103.50    5.0      11.91   3.25           13   13  
798   88.50    5.0       9.93   7.43            8    8  
799   86.80    5.0       6.63   3.20           11   11  
800   88.80    5.0       7.88   4.75           10   10  

[801 rows x 16 columns]

#arrange the column

In [128]:
sorted_data = d3.sort_values(by='Price' , ascending=False)
sorted_data

Unnamed: 0                                         Name    Location  \
531        4079  Land Rover Range Rover 3.0 Diesel LWB Vogue   Hyderabad   
197        1505              Land Rover Range Rover Sport SE       Kochi   
277        2095                     Mercedes-Benz SLC 43 AMG  Coimbatore   
243        1885         Mercedes-Benz GLS 350d Grand Edition       Delhi   
312        2319                Mercedes-Benz E-Class E 350 d  Coimbatore   
..          ...                                          ...         ...   
52          351                             Skoda Octavia RS      Jaipur   
423        3132                         Porsche Cayenne Base       Kochi   
175        1313                     Hyundai Verna CRDi 1.4 E     Chennai   
295        2200                           Maruti Wagon R VXI      Mumbai   
622        4863                        Hyundai Santro D Lite     Chennai   

     original_Year  Kilometers_Driven  Fuel_Type  Transmission  Owner_Type  \
531           2017              25000          0             1           1   
197           2019              26013          0             1           1   
277           2019               2526          1             1           1   
243           2018               6000          0             1           1   
312           2019               9312          0             1           1   
..             ...                ...        ...           ...         ...   
52            2005             100000          1             1           1   
423           2019              14298          1             1           1   
175           2008             121097          0             0           2   
295           2010              38000          1             0           2   
622           2001             140000          1             0           3   

     Mileage  Engine   Power  Seats  New_Price   Price  Current_Age  Age  
531    13.33  2993.0   88.80    5.0       2.30  160.00            7    7  
197    12.65  2993.0   88.80    5.0       1.39   97.07            5    5  
277    19.00  2996.0  362.07    2.0       1.06   83.96            5    5  
243    11.00  2987.0   88.80    7.0       1.02   79.00            6    6  
312    18.00  2987.0   88.80    5.0      92.50   78.80            5    5  
..       ...     ...     ...    ...        ...     ...          ...  ...  
52     14.45  1984.0   88.80    5.0      28.37    2.05           19   19  
423    13.33  2995.0   88.80    5.0       1.36    2.02            5    5  
175    24.00  1396.0   88.76    5.0      10.73    1.90           16   16  
295    22.50   998.0   67.04    5.0       5.61    1.75           14   14  
622    20.30  1086.0   88.80    5.0       4.55    0.70           23   23  

[801 rows x 16 columns]

# Summarize

In [129]:
average_price_by_make = d3.groupby('Seats')['Price'].mean()
average_price_by_make

Seats
2.0    68.980000
4.0    34.555833
5.0    12.515426
6.0     7.665000
7.0    17.211356
8.0    18.655714
Name: Price, dtype: float64

In [130]:
d3.head()

Unnamed: 0                         Name Location  original_Year  \
0           2                 Honda Jazz V  Chennai           2011   
1          10             Maruti Ciaz Zeta    Kochi           2018   
2          15  Mitsubishi Pajero Sport 4X4    Delhi           2014   
3          20            BMW 3 Series 320d    Kochi           2014   
4          28          Honda WRV i-VTEC VX    Kochi           2018   

   Kilometers_Driven  Fuel_Type  Transmission  Owner_Type  Mileage  Engine  \
0              46000          1             0           1    18.19  1199.0   
1              25692          1             0           1    21.56  1462.0   
2             110000          0             0           1    13.50  2477.0   
3              32982          0             1           1    22.69  1995.0   
4              37430          1             0           1    17.50  1199.0   

    Power  Seats  New_Price  Price  Current_Age  Age  
0   88.70    5.0       8.61   4.50           13   13  
1  103.25    5.0      10.65   9.95            6    6  
2  175.56    7.0      32.01  15.00           10   10  
3   88.80    5.0      47.87  18.55           10   10  
4   88.70    5.0      10.57   9.90            6    6